vscode-dotnet-runtime-library/src/Utils/WebRequestWorker.ts

```diff
@@ -2,57 +2,164 @@
  *  Copyright (c) Microsoft Corporation. All rights reserved.
  *  Licensed under the MIT License. See License.txt in the project root for license information.
  *--------------------------------------------------------------------------------------------*/
-import retry from 'p-retry';
-import * as request from 'request-promise-native';
+import Axios from 'axios';
+import axiosRetry from 'axios-retry';
+import { AxiosCacheInstance, buildStorage, setupCache, StorageValue } from 'axios-cache-interceptor';
 import { IEventStream } from '../EventStream/EventStream';
 import { WebRequestError, WebRequestSent } from '../EventStream/EventStreamEvents';
 import { IExtensionState } from '../IExtensionState';
+import { Debugging } from '../Utils/Debugging';
+
+/*
+This wraps the VSCode memento state blob into an axios-cache-interceptor-compatible Storage.
+(The momento state is used to save extensionState/data across runs of the extension.)
+All the calls are synchronous.
+*/
+const mementoStorage = (extensionStorage: IExtensionState) => {
+    const cachePrefix = 'axios-cache'; // Used to make it easier to tell what part of the extension state is from the cache
+    return buildStorage({
+        // tslint:disable-next-line
+        set(key: string, value: any) {
+            extensionStorage.update(`${cachePrefix}:${key}`, value);
+        },
+        remove(key: string) {
+            extensionStorage.update(`${cachePrefix}:${key}`, undefined);
+        },
+        find(key: string) {
+            return extensionStorage.get(`${cachePrefix}:${key}`) as StorageValue;
+        }
+    });
+}
 
 export class WebRequestWorker {
-    private cachedData: string | undefined;
-    private currentRequest: Promise<string | undefined> | undefined;
+    /**
+     * @remarks
+     * An interface for sending get requests to APIS.
+     * The responses from GET requests are cached with a 'time-to-live' of 5 minutes by default.
+     */
+    private client: AxiosCacheInstance;
 
-    constructor(private readonly extensionState: IExtensionState,
-                private readonly eventStream: IEventStream,
-                private readonly url: string,
-                private readonly extensionStateKey: string) {}
+    constructor(
+        private readonly extensionState: IExtensionState,
+        private readonly eventStream: IEventStream,
+        private readonly url: string,
+        private readonly websiteTimeoutMs: number, // Match the default timeout time of 10 minutes.
+        private cacheTimeToLive = -1
+        )
+        {
+            this.cacheTimeToLive = this.cacheTimeToLive === -1 ? this.websiteTimeoutMs * 5 : this.cacheTimeToLive; // make things live 5x (arbitrary) as long as it takes to maximally acquire them
+            const uncachedAxiosClient = Axios.create({});
```

Comment:
Why are we using create without setting the timeout and such? It looks like we set it below, so I guess it's not a big deal

Reply:
I think I did not want to make it specific to one url, I could go either way on this. Will leave it for now